core/execution/future/market.go

```diff
@@ -997,6 +997,67 @@ func (m *Market) parkAllPeggedOrders(ctx context.Context) []*types.Order {
 	return parked
 }
 
+func (m *Market) uncrossOrderAtAuctionEnd(ctx context.Context) {
+	if !m.as.InAuction() {
+		return
+	}
+	// Change market type to continuous trading
+	uncrossedOrders, _, err := m.matching.LeaveAuction(m.timeService.GetTimeNow())
+	if err != nil {
+		m.log.Error("Error leaving auction", logging.Error(err))
+	}
+
+	// Process each confirmation & apply fee calculations to each trade
+	evts := make([]events.Event, 0, len(uncrossedOrders))
+	for _, uncrossedOrder := range uncrossedOrders {
+		// handle fees first
+		err := m.applyFees(ctx, uncrossedOrder.Order, uncrossedOrder.Trades)
+		if err != nil {
+			// @TODO this ought to be an event
+			m.log.Error("Unable to apply fees to order",
+				logging.String("OrderID", uncrossedOrder.Order.ID))
+		}
+
+		// then do the confirmation
+		m.handleConfirmation(ctx, uncrossedOrder)
+
+		if uncrossedOrder.Order.Remaining == 0 {
+			uncrossedOrder.Order.Status = types.OrderStatusFilled
+		}
+		evts = append(evts, events.NewOrderEvent(ctx, uncrossedOrder.Order))
+	}
+
+	// send order events in a single batch, it's more efficient
+	m.broker.SendBatch(evts)
+}
+
+func (m *Market) UpdateMarketState(ctx context.Context, changes *types.MarketStateUpdateConfiguration) error {
+	if changes.UpdateType == types.MarketStateUpdateTypeTerminate {
+		m.uncrossOrderAtAuctionEnd(ctx)
+		// terminate and settle
+		n := num.Numeric{}
+		n.SetUint(changes.SettlementPrice)
+		m.settlementData(ctx, &n)
+		// first we set the settlement price so it is used then we terminate the market.
+		m.tradingTerminatedWithFinalState(ctx, types.MarketStateClosed)
+	} else if changes.UpdateType == types.MarketStateUpdateTypeSuspend {
+		m.mkt.State = types.MarketStateSuspendedViaGovernance
+		m.mkt.TradingMode = types.MarketTradingModeSuspendedViaGovernance
+		m.as.StartGovernanceSuspensionAuction(m.timeService.GetTimeNow())
+		m.enterAuction(ctx)
+		m.broker.Send(events.NewMarketUpdatedEvent(ctx, *m.mkt))
+	} else if changes.UpdateType == types.MarketStateUpdateTypeResume && m.mkt.State == types.MarketStateSuspendedViaGovernance {
+		m.as.EndGovernanceSuspensionAuction()
+		m.checkAuction(ctx, m.timeService.GetTimeNow(), m.idgen)
```

Comment:
I think we need to call `leaveAuction()` somewhere otherwise we won't get an `AuctionState` event saying we've left, and also unpark pegged orders etc etc.

Reply:
Also I can't see anything in `checkAuction` that would re-enter us into an opening-auction if the suspension happened while in one (and then was resumed while still in the period)?

Reply:
indeed, returning to an opening auction is a bit of a problem with this implementation. need to think about it.